src/opf/acp.jl

```diff
@@ -1,293 +1,167 @@
-        sum(qg[g] for g in ref[:bus_gens][i]) 
-        - sum(qf[a] for a in ref[:bus_arcs][i])
-        + sum(shunt["bs"] for shunt in bus_shunts[i])*vm[i]^2
+        sum(gen_status[g] * qg[g] for g in bus_gens[i])
+        - sum(branch_status[e] * qf[e] for e in bus_arcs_fr[i])
+        - sum(branch_status[e] * qt[e] for e in bus_arcs_to[i])
+        + bs[i] * vm[i]^2
         ==
-        sum(load["qd"] for load in bus_loads[i])
+        qd[i]
     )
 
-    # Branch power flow physics and limit constraints
-    # We pre-allocate constraint containers for simplicity
-    model[:thermal_limit_fr] = Vector{JuMP.ConstraintRef}(undef, E)
-    model[:thermal_limit_to] = Vector{JuMP.ConstraintRef}(undef, E)
-    model[:voltage_difference_limit] = Vector{JuMP.ConstraintRef}(undef, E)
-    model[:ohm_active_fr] = Vector{JuMP.ConstraintRef}(undef, E)
-    model[:ohm_active_to] = Vector{JuMP.ConstraintRef}(undef, E)
-    model[:ohm_reactive_fr] = Vector{JuMP.ConstraintRef}(undef, E)
-    model[:ohm_reactive_to] = Vector{JuMP.ConstraintRef}(undef, E)
-    for (i,branch) in ref[:branch]
-        data["branch"]["$i"]["br_status"] == 0 && continue  # skip branch
-
-        f_idx = (i, branch["f_bus"], branch["t_bus"])
-        t_idx = (i, branch["t_bus"], branch["f_bus"])
-
-        p_fr = pf[f_idx]
-        q_fr = qf[f_idx]
-        p_to = pf[t_idx]
-        q_to = qf[t_idx]
-
-        vm_fr = vm[branch["f_bus"]]
-        vm_to = vm[branch["t_bus"]]
-        va_fr = va[branch["f_bus"]]
-        va_to = va[branch["t_bus"]]
-
-        g, b = PM.calc_branch_y(branch)
-        tr, ti = PM.calc_branch_t(branch)
-        ttm = tr^2 + ti^2
-        g_fr = branch["g_fr"]
-        b_fr = branch["b_fr"]
-        g_to = branch["g_to"]
-        b_to = branch["b_to"]
-
-        # From side of the branch flow
-        model[:ohm_active_fr][i] = JuMP.@constraint(model, 
-              ((g+g_fr)/ttm) * vm_fr^2
-            + ((-g*tr+b*ti)/ttm) * (vm_fr*vm_to*cos(va_fr-va_to)) 
-            + ((-b*tr-g*ti)/ttm) * (vm_fr*vm_to*sin(va_fr-va_to)) 
-            - p_fr 
-            ==
-            0
+    model[:sm_fr] = Vector{ConstraintRef}(undef, E)
+    model[:sm_to] = Vector{ConstraintRef}(undef, E)
+    model[:va_diff] = Vector{ConstraintRef}(undef, E)
+    model[:ohm_pf] = Vector{ConstraintRef}(undef, E)
+    model[:ohm_qf] = Vector{ConstraintRef}(undef, E)
+    model[:ohm_pt] = Vector{ConstraintRef}(undef, E)
+    model[:ohm_qt] = Vector{ConstraintRef}(undef, E)
+
+    @expression(model, wf[e in 1:E], vm[bus_fr[e]]^2)
+    @expression(model, wt[e in 1:E], vm[bus_to[e]]^2)
+    @expression(model, wr[e in 1:E], vm[bus_fr[e]] * vm[bus_to[e]] * cos(va[bus_fr[e]] - va[bus_to[e]]))
+    @expression(model, wi[e in 1:E], vm[bus_fr[e]] * vm[bus_to[e]] * sin(va[bus_fr[e]] - va[bus_to[e]]))
+
+    for e in 1:E
+        # Ohm's law
+        model[:ohm_pf][e] = @constraint(model,
+            branch_status[e] * ( gff[e] * wf[e] + gft[e] * wr[e] + bft[e] * wi[e]) - pf[e] == 0
```

Comment:
One small thing that comes to mind: if a branch is not active, I don't know whether JuMP will parse this as `0 == 0` or as a nonlinear expression that happens to have zeros everywhere, e.g., `0 * wf[e] + 0 * wr[e] + 0 * wi[e]  - pf[e] == 0`.

I think the latter would trigger extra computations evaluating gradients & sparsity patterns.
Not a big concern, just a note.

Reply:
Confirmed: JuMP will keep the full expression graph, with zero-valued coefficients. This is a small computational that could be avoided (since all function evaluation and gradients will be zero). Should not be significant as long as we only deactivate a small number of components.